hail/src/main/scala/is/hail/types/physical/PCanonicalArray.scala

```diff
@@ -459,4 +461,48 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
 
   private def deepRenameArray(t: TArray): PArray =
     PCanonicalArray(this.elementType.deepRename(t.elementType), this.required)
+
+
+  def constructFromElements(cb: EmitCodeBuilder, region: Value[Region], length: Value[Int], deepCopy: Boolean)
+    (f: (EmitCodeBuilder, Value[Int]) => IEmitCode): SIndexablePointerCode = {
+
+    val addr = cb.newLocal[Long]("pcarray_construct1_addr", allocate(region, length))
+    cb += stagedInitialize(addr, length, setMissing = false)
+    val i = cb.newLocal[Int]("pcarray_construct1_i", 0)
+
+    val firstElementAddr = firstElementOffset(addr, length)
+    cb.whileLoop(i < length, {
+      f(cb, i).consume(cb,
+        setElementMissing(addr, i),
+        { sc =>
+          elementType.storeAtAddress(cb, elementOffsetFromFirst(firstElementAddr, i), region, sc, deepCopy = deepCopy)
+        })
+
+      cb.assign(i, i + 1)
+    })
+
+    new SIndexablePointerCode(SIndexablePointer(this), addr)
+  }
+
+  // unsafe StagedArrayBuilder-like interface that gives caller control over adding elements and finishing
+  def constructFromFunctions(cb: EmitCodeBuilder, region: Value[Region], length: Value[Int], deepCopy: Boolean):
+  (((EmitCodeBuilder, Value[Int], IEmitCode) => Unit, (EmitCodeBuilder => SIndexablePointerCode))) = {
+
+    val addr = cb.newLocal[Long]("pcarray_construct1_addr", allocate(region, length))
+    cb += stagedInitialize(addr, length, setMissing = false)
+    val i = cb.newLocal[Int]("pcarray_construct1_i", 0)
```

Comment:
Should the name here be `"pcarray_construct2_i"`? I assumed the 1 was to distinguish between these two constructors.

Reply:
oh, yes.